synapse/tools/backup.py

```diff
@@ -0,0 +1,119 @@
+import os
+import sys
+import time
+import shutil
+import logging
+import argparse
+import datetime
+
+import lmdb
+import lmdb.tool
+
+import synapse.common as s_common
+
+logger = logging.getLogger(__name__)
+
+def backup(srcdir, backupdir):
+    tick = s_common.now()
+    srcdir = s_common.reqdir(srcdir)
+    backupdir = s_common.reqdir(backupdir)
+    logger.info(f'Starting backup of [{srcdir}]')
+    logger.info(f'Destination dir: [{backupdir}]')
+
+    srcbase = os.path.dirname(srcdir)
+
+    lmdb_dirs = []
+    for root, dnames, fnames in os.walk(srcdir, topdown=True):
+        relroot = root[len(srcbase) + 1:]
+
+        filter_dirs = []
+        for i, dname in enumerate(dnames):
+            if dname.endswith('.lmdb'):
+                filter_dirs.append(dname)
+                continue
+
+        # separate the lmdb directories
+        for fname in filter_dirs:
+            dnames.remove(fname)
+            src = s_common.genpath(root, fname)
+            dst = s_common.genpath(backupdir, relroot, fname)
+            lmdb_dirs.append((src, dst))
+
+        # make regular directories
+        for dname in dnames:
+            src = s_common.genpath(root, dname)
+            dst = s_common.genpath(backupdir, relroot, dname)
+            logger.info(f'making dir:{dst}')
+            os.makedirs(dst)
+
+        # copy regular files
+        for fname in fnames:
+            src = s_common.genpath(root, fname)
+            dst = s_common.genpath(backupdir, relroot, fname)
+            logger.info(f'copying: {src} -> {dst}')
+            shutil.copy(src, dst)
+
+    # backup each lmdb directory
+    for src, dst in lmdb_dirs:
+        logger.info(f'lmdb backup: {src} -> {dst}')
+        backup_lmdb(src, dst)
+    tock = s_common.now()
+    logger.info(f'Backup complete. Took [{tock-tick:.2f}] for [{srcdir}]')
+    return
+
+def backup_lmdb(envpath, backupdir):
+    # Assumes data.mdb exists.
+    datafile = os.path.join(envpath, 'data.mdb')
```

Comment:
we could try to auto-discover .mdb files; but all our synapse use cases use the lmdb default names for environments

Reply:
I think this is perfectly fine for now as long as it is documented. This is not a general case backup tool, but rather a specialized one to back up a cortex.